File: src/server/src/demo02_client.cpp
#include "ros/ros.h"
#include "server/AddInts.h"


/*
若要实现参数的动态提交：
格式：rosrun xxxx xxx 12 1234
此时 argc为3， argv为 xxx 12 1234 
节点执行时，需要获取命令中的参数，并组织进request
*/


int main(int argc, char *argv[]){

    // 优化实现，获取命令中参数
    if (argc != 3){
        ROS_ERROR("Wrong number of arguments");
        return 1;
    }

    setlocale(LC_ALL, "");
    
    ros::init(argc, argv,"client");
    ros::NodeHandle nh;

    ros::ServiceClient client = nh.serviceClient<server::AddInts>("addInts");

    //提交请求并处理响应
    server::AddInts ai;
    //组织请求
    ai.request.num1 = atoi(argv[1]);
    ai.request.num2 = atoi(argv[2]);

    //处理响应
    bool flag = client.call(ai);
    if (flag) {
        ROS_INFO("Sum: %d", ai.response.sum);
    } else {
        ROS_ERROR("Failed to call service AddInts");
    }

    ros::spin();
    return 0;
}

File: src/server/scripts/demo02_client.py
#! /usr/bin/env python
import rospy
from server.srv import AddInts,AddIntsRequest,AddIntsResponse
import sys
def main():
    """
    客户端：组织并提交请求，处理服务端响应。

    优化实现：可以在执行节点时，动态传入参数
    """
    # 首先判断参数长度
    if len(sys.argv) != 3 :
        rospy.logerror("wrong number of arguments")
        sys.exit(1)
    
    # 解析并获取参数
    num1 = int(sys.argv[1])
    num2 = int(sys.argv[2])

    rospy.init_node("client")
    # 创建客户端对象
    client = rospy.ServiceProxy("addInts",AddInts)
    # 组织请求数据，并发布请求

    """
    解决客户端先于服务端启动
    """
    # 1. 等待服务器启动
    # client.wait_for_service()

    # or
    rospy.wait_for_service("addInts")
    response = client.call(num1,num2)

    # 处理并打印服务端的响应
    rospy.loginfo(f"客户端得到的结果: sum {response.sum}")

    rospy.spin()


if __name__ == "__main__":
    main()